storage-dynamodb/src/main/java/zipkin2/storage/dynamodb/DynamoDBSpanConsumer.java

```diff
@@ -0,0 +1,269 @@
+import static zipkin2.storage.dynamodb.DynamoDBConstants.Spans.LOCAL_SERVICE_NAME;
+import static zipkin2.storage.dynamodb.DynamoDBConstants.Spans.LOCAL_SERVICE_SPAN_NAME;
+import static zipkin2.storage.dynamodb.DynamoDBConstants.Spans.REMOTE_SERVICE_NAME;
+import static zipkin2.storage.dynamodb.DynamoDBConstants.Spans.REMOTE_SERVICE_SPAN_NAME;
+import static zipkin2.storage.dynamodb.DynamoDBConstants.Spans.SPAN_BLOB;
+import static zipkin2.storage.dynamodb.DynamoDBConstants.Spans.SPAN_DURATION;
+import static zipkin2.storage.dynamodb.DynamoDBConstants.Spans.SPAN_ID;
+import static zipkin2.storage.dynamodb.DynamoDBConstants.Spans.SPAN_NAME;
+import static zipkin2.storage.dynamodb.DynamoDBConstants.Spans.SPAN_TIMESTAMP;
+import static zipkin2.storage.dynamodb.DynamoDBConstants.Spans.TAG_PREFIX;
+import static zipkin2.storage.dynamodb.DynamoDBConstants.Spans.SPAN_TIMESTAMP_ID;
+import static zipkin2.storage.dynamodb.DynamoDBConstants.Spans.TRACE_ID;
+import static zipkin2.storage.dynamodb.DynamoDBConstants.Spans.TRACE_ID_64;
+import static zipkin2.storage.dynamodb.DynamoDBConstants.TTL_COLUMN;
+import static zipkin2.storage.dynamodb.DynamoDBConstants.WILDCARD_FOR_INVERTED_INDEX_LOOKUP;
+
+final class DynamoDBSpanConsumer implements SpanConsumer {
+  private static Random random = new Random();
+
+  private final List<String> autocompleteKeys;
+  private final AmazonDynamoDB dynamoDB;
+  private final long dataTtlSeconds;
+  private final String spansTableName;
+  private final String searchTableName;
+
+  DynamoDBSpanConsumer(DynamoDBStorage.Builder builder) {
+    this.autocompleteKeys = builder.autocompleteKeys;
+    this.dynamoDB = builder.dynamoDB;
+    this.dataTtlSeconds = builder.dataTtl.toMillis() / 1000;
+
+    this.spansTableName = builder.tablePrefix + SPANS_TABLE_BASE_NAME;
+    this.searchTableName = builder.tablePrefix + SEARCH_TABLE_BASE_NAME;
+  }
+
+  @Override public Call<Void> accept(List<Span> list) {
+    AttributeValue ttlForBatch = new AttributeValue().withN(
+        String.valueOf(Instant.now().getEpochSecond() + dataTtlSeconds));
+
+    List<WriteRequest> spanWriteRequests = createWriteSpans(list, ttlForBatch);
+    while (!spanWriteRequests.isEmpty()) {
+      int maxIndex = Math.min(spanWriteRequests.size(), 25);
+      BatchWriteItemRequest request = new BatchWriteItemRequest();
+      request.addRequestItemsEntry(spansTableName, spanWriteRequests.subList(0, maxIndex));
+
+      dynamoDB.batchWriteItem(request);
+      spanWriteRequests.subList(0, maxIndex).clear();
+    }
+
+    // We use update requests for the next two entities because we want to update the TTL if they
+    // exist already. Updates in DynamoDB are treated as upserts
+    List<UpdateItemRequest> upserts = createUpsertsForServiceSpanNames(list, ttlForBatch);
+    upserts.addAll(createUpsertsForAutocompleteTags(list, ttlForBatch));
+
+    for (UpdateItemRequest request : upserts) {
+      dynamoDB.updateItem(request.withTableName(searchTableName));
```

Comment:
It seems that Zipkin supports async calls, but you're not using DynamoDB's async client.
Any reason why? Just curious :)

Reply:
in chat @devinsba mentioned his first go was making things pass tests. I've some work locally using the async api and will push later

Reply:
Did you use AWS SDK 2? Version 1 is much harder to deal with (lacking some neat features like pagination, `CompletableFuture` return types, etc) and also caused some memory leaks in one of my projects which we fixed by [rewriting to AWS SDK 2](https://github.com/bsideup/liiklus/pull/77).

Reply:
Yeah, I realized after the fact. It shouldn't be that bad to swap out but it's worth discussing

Reply:
That does bring up the potential for classpath issues though. We couldn't do netty I'm pretty sure. Maybe we could do Apache HttpClient but I know that isn't a favorite choice

Reply:
AFAIR it is pluggable and you can bring your own http transport layer, already being used by Zipkin

Reply:
Yeah, the existing impls now though are Apache, Netty, and UrlConnection. Would be nice to have OkHttp (or some other minimal dep client) for sure as Adrian requested a long while ago

Reply:
async is done now, pagination another beast